src/extra/recipes/cookiebox.py

```diff
@@ -0,0 +1,1047 @@
+from lmfit.model import ModelResult
+import scipy
+from scipy.stats import linregress
+from scipy.optimize import minimize_scalar
+from functools import partial
+from scipy.signal import fftconvolve
+from scipy.interpolate import CubicSpline
+from scipy.signal import kaiserord, filtfilt, firwin
+
+
+@dataclass
+class TofFitResult:
+    """Class keeping track of fit in a single Tof."""
+    energy: np.ndarray
+    mu: np.ndarray
+    sigma: np.ndarray
+    A: np.ndarray
+    Aa: np.ndarray
+    offset: np.ndarray
+
+
+def search_offset(trace: np.ndarray, sigma: float=20):
+    """
+    Find highest peaks in the 1D trace.
+    """
+    axis = np.arange(len(trace))
+    gaussian = np.exp(-0.5*(axis - np.mean(axis))**2/sigma**2)
+    gaussian /= np.sum(gaussian, axis=0, keepdims=True)
+    # apply it to the data
+    smoothened = fftconvolve(trace, gaussian, mode="same", axes=0)
+    peak_idx = np.argmax(smoothened)
+    return peak_idx - 200
+
+def search_roi(roi: np.ndarray):
+    """
+    Find highest peaks in the 1D trace.
+    """
+    p, _ = scipy.signal.find_peaks(roi, prominence=(0.25*np.amax(roi), None))
+    return p
+
+def model(ts,c,e0,t0):
+    """
+    Model to fit energy as a function of the sample axis.
+    """
+    # e = e0 + c/(ts - t0)**2
+    # ts = np.sqrt(c/(e - e0)) + t0
+    # dt/de = d/de [ sqrt(c/e) ] = d/du sqrt(u) d/de [c/e] = -0.5*c/sqrt(c/e) * 1/(e-e0)^2 
+    # dt/de = -0.5*c/(np.sqrt(c/(e - e0)))/(e - e0)**2
+    # before transmission, after time -> energy interpolation:
+    # multiply peak sizes by dt/de
+    return e0+c/(ts-t0)**2
+
+# e = e0 + c/(ts-t0)^2
+# 1/(e - e0)
+def lin_fit(ts,energies,t0):
+    """
+    Fit model using linear regression for fixed t0.
+    """
+    res = linregress(1/np.asarray(ts-t0)**2,np.asarray(energies))
+    c = res.slope
+    e0 = res.intercept
+    return c,e0,t0
+
+def norm_diff_t0(ts,energies,t0):
+    #print(ts,'fi',energies)
+    c,e0,_ = lin_fit(ts,energies,t0)
+    return np.linalg.norm(model(ts,c,e0,t0) - energies)
+
+def fit(peak_ids,energies,t0_bounds):
+    peak_ids = np.asarray(peak_ids)
+    energies = np.asarray(energies)
+    #print(peak_ids)
+    #print(energies)
+
+    f = partial(norm_diff_t0,peak_ids,energies)
+    res=minimize_scalar(f,method='Bounded',bounds=t0_bounds)
+    t0 = res.x
+    #print(res)
+    c,e0,t0=lin_fit(peak_ids,energies,t0)
+    if res.success:
+        return c,e0,t0
+    else:
+        raise Exception('fit did not converge.')
+
+
+def calc_mean(itr: Tuple[int, int], scan: Scan, xgm_data: xr.DataArray, tof: Dict[int, AdqRawChannel], xgm_threshold: float, filter_length: int) -> xr.DataArray:
+    """A function doing the actual calibration, per tof, per energy index."""
+    tof_id, energy_id = itr
+    energy, train_ids = scan.steps[energy_id]
+    #xgm = XGM(run.select_trains(by_id[list(train_ids)]), "SA3_XTD10_XGM/XGM/DOOCS")
+    #sel_xgm_data = xgm.pulse_energy().stack(pulse=('trainId', 'pulseIndex'))
```

Comment:
Nitpicking, I think this can be deleted.

Reply:
Agreed.